File: Ryujiho/LeetCode/3_Longest Substring Without Repeating Characters_String.py
class Solution(object):

  def lengthOfLongestSubstring(self, s):
    maxLen = 0
    q = []
    for i in s:
      if i not in q:
        q.append(i)
      else:
        idx = q.index(i)
        del q[:idx + 1]
        q.append(i)

      maxLen = max(maxLen, len(q))

    return maxLen

File: eileenjang/LeetCode/3_Longest Substring Without Repeating Characters_String.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        int index = 0;
        int max = 0;
        int prev = 0;

        Set<Character> set = new HashSet<>();
        
        while (index < s.length()) {
            char c = s.charAt(index);
            if(!set.contains(c)) {
                set.add(c);
                index++;
                max = Math.max(set.size(), max);
            }
            else {
                set.remove(s.charAt(prev));
                prev++;
            }
        }
        return max;
    }
}
